File: framework/scala/src/main/scala/soda/scala/leetcode/ListFactory.scala
package soda.scala.leetcode

import scala.collection.mutable.ArrayBuffer

object ListFactory {

  def create(values: List[Int]): ListNode = {
    val head = new ListNode()
    var tail = head
    for (value <- values) {
      val node = new ListNode(value)
      tail.next = node
      tail = node
    }
    head.next
  }

  def dump(head: ListNode): List[Int] = {
    var p = head
    val list = ArrayBuffer[Int]()
    while (p != null) {
      list += p.x
      p = p.next
    }
    list.toList
  }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/leetcode/ListFactory.kt
package soda.kotlin.leetcode

object ListFactory {

    fun create(values: List<Int>): ListNode? {
        val head = ListNode(-1)
        var tail = head
        for (value in values) {
            val node = ListNode(value)
            tail.next = node
            tail = node
        }
        return head.next
    }

    fun dump(head: ListNode?): List<Int> {
        var p = head
        val list = mutableListOf<Int>()
        while (p != null) {
            list += p.`val`
            p = p.next
        }
        return list.toList()
    }

}